### ISSUE TITLE
Prototype single repo

### ISSUE THREAD

**Issue by jw3** [jw3/example-daffodil-vscode#37]
Scala & Extension in harmony

**Comment by arosien** [comment]
I like this idea, especially because it lessens the requirements for releasing new versions of either the backend or frontend: they would be the same thing. It would also remove the "choose a backend version" dialog on launch.

**Comment by jw3** [comment]
It would indeed eliminate the need to pick a different DAP backend, but it would be nice to keep the ability to run against multiple versions of Daffodil from a single extension.

The "choose a backend" dialog might become a choose a version of Daffodil dialog.

**Comment by kirkejohnson** [comment]
I think Mike and Steve would appreciate that option.  "Choose the version of Daffodil..."

**Comment by arosien** [comment]
To support multiple Daffodil versions in the debugger, assuming Daffodil binary compatibility, we would need to:

* Mark the Daffodil dependencies as `provided`. This would compile the debugger against the libraries, but not include them in the artifact. The current dependencies are `"org.apache.daffodil" %% "daffodil-sapi" % daffodilVer` and `  "org.apache.daffodil" %% "daffodil-runtime1" % daffodilVer`.
* Download (and cache) Daffodil releases, for example, https://www.apache.org/dyn/closer.lua/daffodil/3.1.0/bin/apache-daffodil-3.1.0-bin.tgz. These releases must include Daffodil and all transitive dependencies needed by Daffodil. Unpack somewhere.

What I don't know is how to add the Daffodil dependencies to the classpath of the packaged debugger. Looking at the `da-podil` script, the classpath is hard-coded and doesn't seem to be able to be modified.

**Comment by arosien** [comment]
> What I don't know is how to add the Daffodil dependencies to the classpath of the packaged debugger. Looking at the da-podil script, the classpath is hard-coded and doesn't seem to be able to be modified.

To "link" multiple artifacts at runtime is a tough coordination problem, and the tooling discourages it. So I think to support multiple Daffodil versions, we'll need to release multiple debugger artifacts for each debugger release.

**Comment by jw3** [comment]
I was thinking of https://www.scala-sbt.org/1.x/docs/Howto-Customizing-Paths.html#Change+the+default+%28unmanaged%29+library+directory we we originally discussed this.

**Comment by jw3** [comment]
> So I think to support multiple Daffodil versions, we'll need to release multiple debugger artifacts for each debugger release.

This scales fine for 0.1.0.

**Comment by arosien** [comment]
> I was thinking of https://www.scala-sbt.org/1.x/docs/Howto-Customizing-Paths.html#Change+the+default+%28unmanaged%29+library+directory we we originally discussed this.

I believe that only applies to the build process, but when you package using sbt-native-packager, all the libraries need to be present so they are included in the artifact. The artifact it builds can't, afaict, have a classpath that is extended at runtime (choosing the Daffodil version on the user's machine).

**Comment by jw3** [comment]
Yeah, that was my impression after looking into it this morning.  So I think what you said above is good for our 0.1.0.

> The artifact it builds can't, afaict, have a classpath that is extended at runtime (choosing the Daffodil version on the user's machine).

I think one could still build against the Daffodil API as a provided artifact, exclude it during packaging, and do some post processing of the script that native packager generates to add the version specific path to the classpath.

That's an enhancement in my mind, not _required_ for right now.

Anyway 0.1.0 will still support 100% of the releases possible right now.

**Comment by arosien** [comment]
Ok, this is super weird: I have no idea what causes the `example-daffodil-debug` packaging to name the launch script `da-podil.sh`.

With the Scala code migrated to the `example-daffodil-vscode` repo, the script name is `daffodil-debugger`, which corresponds to the configured name of the sbt project (`name := "daffodil-debugger"`).

I suppose this is ok, it's just a mystery for the prior repo.

**Comment by arosien** [comment]
I'm guessing the prior script naming is because of this:

```
sbt:daffodil-debugger> show discoveredMainClasses
[info] * ddb.debugger.dapodil.DAPodil
[info] * ddb.debugger.z.main
```

And somehow `DAPodil` gets escaped by the packaging plugin to `da-podil`...?

**Comment by jw3** [comment]
Yeah, looks like a combination of multiple main methods detected and some camel-to-lower logic.

Looking at this I was surprised the source was merged together.  Was expecting a separate dirs for scala and typescript.

Is this how other extensions do it?

**Comment by arosien** [comment]
After experimenting with the wrong thing a couple of times, I think the extension changes for a monorepo need to be:

* Build actions should additionally build the backend artifact and include it in any releases.
* We don't want the user to pick the version of the backend, the backend should be co-released with the extension. So don't query for the tags of the project repository. The backend release artifact is directly resolvable from the version of the extension, i.e., extension 0.0.10 maps to `https://github.com/${backend.owner}/${backend.repo}/releases/download/v0.0.10/daffodil-debugger-0.0.10.zip`. I'm not sure how to get the extension version into the extension source code, but [this](https://stackoverflow.com/questions/41123631/how-to-get-the-version-from-the-package-json-in-typescript/67701490#67701490) seems to be one idea.
* In the future, we'll want the user to choose Daffodil versions associated with the specific extension+backend version. To support that, we'd add the Daffodil version to the name of the backend artifact, e.g., `daffodfil-debugger-3.1.0-0.0.10.zip` vs. `daffodfil-debugger-3.2.0-0.0.10.zip`, etc.

**Comment by jw3** [comment]
closed by #76
